src/main/java/com/o19s/es/ltr/ranker/parser/XGBoostJsonParser.java

```diff
@@ -41,60 +44,109 @@
         }
-        float[] weights = new float[trees.size()];
+
+        Node[] trees = modelDefinition.getTrees(set);
+        float[] weights = new float[trees.length];
         // Tree weights are already encoded in outputs
         Arrays.fill(weights, 1F);
-        return new NaiveAdditiveDecisionTree(trees.toArray(new Node[0]), weights, set.size(), params.objective);
+        return new NaiveAdditiveDecisionTree(trees, weights, set.size(), modelDefinition.normalizer);
     }
 
-    private static class XGBoostParams {
-        private static final ObjectParser<XGBoostParams, Object> PARSER;
+    private static class XGBoostDefinition {
+        private static final ObjectParser<XGBoostDefinition, FeatureSet> PARSER;
         static {
-            PARSER = new ObjectParser<>("params", XGBoostParams::new);
-            PARSER.declareString(XGBoostParams::setObjective, new ParseField("objective"));
+            PARSER = new ObjectParser<>("xgboost_definition", XGBoostDefinition::new);
+            PARSER.declareString(XGBoostDefinition::setNormalizer, new ParseField("objective"));
+            PARSER.declareObjectArray(XGBoostDefinition::setSplitParserStates, SplitParserState::parse, new ParseField("splits"));
         }
 
-        private ModelObjective objective;
+        private Normalizer normalizer;
+        private List<SplitParserState> splitParserStates;
+
+        public static XGBoostDefinition parse(XContentParser parser, FeatureSet set) throws IOException {
+            XGBoostDefinition definition;
+            XContentParser.Token startToken = parser.nextToken();
 
-        public static XGBoostParams parse(XContentParser parser, Object o) {
-            return PARSER.apply(parser, o);
+            // The model definition can either be an array of tree definitions, or an object containing the
+            // tree definitions in the 'splits' field. Using an object allows for specification of additional
+            // parameters.
+            if (startToken == XContentParser.Token.START_OBJECT) {
+                try {
+                    definition = PARSER.apply(parser, set);
+                } catch (XContentParseException e) {
+                    throw new ParsingException(parser.getTokenLocation(), "Unable to parse XGBoost object", e);
+                }
+            } else if (startToken == XContentParser.Token.START_ARRAY) {
+                definition = new XGBoostDefinition();
+                definition.splitParserStates = new ArrayList<>();
+                while (parser.nextToken() != XContentParser.Token.END_ARRAY) {
+                    definition.splitParserStates.add(SplitParserState.parse(parser, set));
+                }
+            } else {
+                throw new ParsingException(parser.getTokenLocation(), "Expected [START_ARRAY] or [START_OBJECT] but got ["
+                        + startToken + "]");
+            }
+            if (!definition.hasAllFields()) {
```

Comment:
nit: perhaps move this check in the `if (startToken == XContentParser.Token.START_OBJECT) {` and inline the `definition.splitParserStates != null` check so that you can inform in he exception message that the field "splits" that is actually missing.

Reply:
I had based this off the split parsing validation. But I guess we can make it more specific for now and reassess if the parsing gets more complicated.